cmake/Modules/FindMKL.cmake

```diff
@@ -1,120 +1,205 @@
+  else()
+    set(INTEL_ROOT_DEFAULT "/opt/intel")
+  endif()
+  set(INTEL_ROOT ${INTEL_ROOT_DEFAULT} CACHE PATH "Folder contains intel libs")
+  find_path(MKL_ROOT include/mkl.h PATHS $ENV{MKLROOT} ${INTEL_ROOT}/mkl
+                                     DOC "Folder contains MKL")
+
+  # ---[ Find include dir
+  find_path(MKL_INCLUDE_DIR mkl.h PATHS ${MKL_ROOT} PATH_SUFFIXES include)
+  set(__looked_for MKL_INCLUDE_DIR)
+
+  # ---[ Find libraries
+  if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+    set(__path_suffixes lib lib/ia32)
+  else()
+    set(__path_suffixes lib lib/intel64)
+  endif()
+
+  set(__mkl_libs "")
+  if(MKL_USE_SINGLE_DYNAMIC_LIBRARY)
+    list(APPEND __mkl_libs rt)
+  else()
+    if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+      if(WIN32)
+        list(APPEND __mkl_libs intel_c)
+      else()
+        list(APPEND __mkl_libs intel gf)
+      endif()
+    else()
+      list(APPEND __mkl_libs intel_lp64 gf_lp64)
+    endif()
+
+    if(MKL_MULTI_THREADED)
+      list(APPEND __mkl_libs intel_thread)
+    else()
+       list(APPEND __mkl_libs sequential)
+    endif()
+
+    list(APPEND __mkl_libs core cdft_core)
+  endif()
+
+  foreach (__lib ${__mkl_libs})
+    set(__mkl_lib "mkl_${__lib}")
+    string(TOUPPER ${__mkl_lib} __mkl_lib_upper)
+
+    if(MKL_USE_STATIC_LIBS)
+      set(__mkl_lib "lib${__mkl_lib}.a")
+    endif()
+
+    find_library(${__mkl_lib_upper}_LIBRARY
+          NAMES ${__mkl_lib}
+          PATHS ${MKL_ROOT} "${MKL_INCLUDE_DIR}/.."
+          PATH_SUFFIXES ${__path_suffixes}
+          DOC "The path to Intel(R) MKL ${__mkl_lib} library")
+    mark_as_advanced(${__mkl_lib_upper}_LIBRARY)
+
+    list(APPEND __looked_for ${__mkl_lib_upper}_LIBRARY)
+    list(APPEND MKL_LIBRARIES ${${__mkl_lib_upper}_LIBRARY})
+  endforeach()
+
+  if(NOT MKL_USE_SINGLE_DYNAMIC_LIBRARY)
+    if (MKL_USE_STATIC_LIBS)
+      set(__iomp5_libs iomp5 libiomp5mt.lib)
+    else()
+      set(__iomp5_libs iomp5 libiomp5md.lib)
+    endif()
+
+    if(WIN32)
+      find_path(INTEL_INCLUDE_DIR omp.h PATHS ${INTEL_ROOT} PATH_SUFFIXES include)
+      list(APPEND __looked_for INTEL_INCLUDE_DIR)
+    endif()
+
+    find_library(MKL_RTL_LIBRARY ${__iomp5_libs}
+       PATHS ${INTEL_RTL_ROOT} ${INTEL_ROOT}/compiler ${MKL_ROOT}/.. ${MKL_ROOT}/../compiler
+       PATH_SUFFIXES ${__path_suffixes}
+       DOC "Path to OpenMP runtime library")
+
+    list(APPEND __looked_for MKL_RTL_LIBRARY)
+    list(APPEND MKL_LIBRARIES ${MKL_RTL_LIBRARY})
+  endif()
+
+  include(FindPackageHandleStandardArgs)
+  find_package_handle_standard_args(MKL DEFAULT_MSG ${__looked_for})
+
+  if(MKL_FOUND)
+    set(CAFFE2_USE_MKL 1)
+    message(STATUS "Found MKL (include: ${MKL_INCLUDE_DIR}, lib: ${MKL_LIBRARIES})")
+  endif()
+
+  caffe_clear_vars(__looked_for __mkl_libs __path_suffixes __lib_suffix __iomp5_libs)
+
+endif()
+
+if(IDEEP_FOUND)
+  set(MKL_FOUND True)
+  list(APPEND MKL_INCLUDE_DIR ${IDEEP_INCLUDE_DIR})
+  list(APPEND MKL_LIBRARIES ${IDEEP_LIBRARIES})
+endif()
+
+set(USE_MKL OFF)
```

Comment:
I feel this part of the logic is overly complicated. Either `IDEEP` or `MKLML` should qualify `USE_ML` as `ON`. (They don't conflict with each other, right?) Then, in the summary part, in addition to showing whether `USE_MKL` is ON or OFF, we should also show whether `CAFFE2_USE_IDEEP` and/or `CAFFE2_USE_MKLML` are on. 

Since we have close sourced `MKLML` disguised as `MKLDNN`, right now. I think show explicit what we are using can help reducing the confusion. Of course, if things are going well, we can gradually deprecate `MKLML`.

Reply:
Yes, they are totally independent.  OK. USE_MKL for existing MKLML, and USE_IDEEP for new now.
Is that OK?

Reply:
Yes, it's OK.